### ISSUE TITLE
SyntaxNodeAction is invoked twice for the same syntax node in case of positional records

### ISSUE THREAD

**Issue by csaba-sagi-sonarsource** [dotnet/roslyn#53136]
Microsoft.CodeAnalysis.CSharp: 3.8.0-5.final 

**Steps to Reproduce**:

1. `RegisterSyntaxNodeAction` for `SyntaxKind.RecordDeclaration`
2. Verify that for the following syntax the action is invoked twice:
 `public record Person(string FirstName, string LastName);`

 Looks like the only difference is in the `ContainingSymbol`.

**Expected Behavior**:
 The action is invoked only once for the RecordDeclaration.
**Actual Behavior**:
The action is invoked twice for the same RecordDeclaration.

**Comment by Youssef1313** [comment]
Duplicate of https://github.com/dotnet/roslyn/issues/50989.

**Comment by Youssef1313** [comment]
@AlekseyTs Could this be related to #46308?

**Comment by Youssef1313** [comment]
That makes sense. But in case the analysis are purely syntactic and the analyzer doesn't check `ContainingSymbol`, it will produce the same diagnostic twice. Is it the responsibility of analyzer author to handle that, or should there be a deduping mechanism in the analyzer driver/executor?

**Comment by CyrusNajmabadi** [comment]
This feels like a bug to me. If I register a syntax action, it should only ever be called once per node.  I don't get how it why the symbolic side would even come into play here. @mavasani ?

**Comment by AlekseyTs** [comment]
Since symbol is part of the context, there is no such thing as "purely syntactic" analyzer, in my opinion.

**Comment by AlekseyTs** [comment]
Quite possible there is a design flaw in the context for a SyntaxNodeAction.

**Comment by CyrusNajmabadi** [comment]
Note: I can think of a single case where this happens with any other nodes. To that extent, as this is a syntax operation, and there is truly only one syntax node with this kind, this feels completely like a bug.  I would say this is also true of the rest of the analyzer callback apis. They should never be called twice with the same 'entity' (be that node, symbol, operation or whatever). No clients expect that, and it will just lead to duplicate diagnostics (and other bugs).

**Comment by Youssef1313** [comment]
I'd say `SyntaxNodeAnalysisContext.ContainingSymbol` shouldn't have been an API. Is it heavily used? If not, is it an option to obsolete it?

**Comment by CyrusNajmabadi** [comment]
I'm confused. Are you saying this is the behavior we should be having?  I fundamentally disagree. If I ask to hear about a syntax kind, then I should be called back once for each nice matching that syntax kind.

**Comment by CyrusNajmabadi** [comment]
> I'd say SyntaxNodeAnalysisContext.ContainingSymbol shouldn't have been an API

It's fine for that to be an api imo. It just needs will defined semantics.

**Comment by CyrusNajmabadi** [comment]
> analyzers should adapt when old behavior is not applicable.

The old behavior is applicable. I argue that it is also desirable. And it should be what all our callbacks attempt to restrict themselves to as an invariant. 

I'm happy to take this to compat council to discuss. However, this behavior is both surprising and inconsistent and didn't seem to be helpful at all. It comes across as a big we should fix, not a 'wat' that consumers sound have to now understand and incorporate.

**Comment by mavasani** [comment]
I agree with @CyrusNajmabadi that there should be a single callback for each syntax node. Given we now have a new concept of same syntax declaring multiple symbols, we should consider adding a new property say `AdditionalContainingSymbols` or some such to accommodate multiple declared symbols.

**Comment by CyrusNajmabadi** [comment]
FWIW, it's a little interesting that we have zero (non-test) usages of .ContainingSymbol that i can find.

**Comment by AlekseyTs** [comment]
@jinujoseph I believe IDE team owns the general analyzers infrastructure now, the AnalyzerDriver, the code under DiagnosticAnalyzer, etc. The code that drives actions is in this area. At this point, I think IDE team should make the call what behavior changes should be made and what API changes, if any, should be made and make the changes.
